netket/optimizer/qgt/qgt_jacobian_pytree.py

```diff
@@ -0,0 +1,196 @@
+
+    @jax.jit
+    def __matmul__(self, vec: Union[PyTree, Array]) -> Union[PyTree, Array]:
+        # Turn vector RHS into PyTree
+        if hasattr(vec, "ndim"):
+            _, unravel = nkjax.tree_ravel(self.params)
+            vec = unravel(vec)
+            ravel = True
+        else:
+            ravel = False
+
+        # Real-imaginary split RHS in R→R and R→C modes
+        if self.mode != "holomorphic":
+            vec, reassemble = nkjax.tree_to_real(vec)
+
+        if self.scale is not None:
+            vec = jax.tree_multimap(jnp.multiply, vec, self.scale)
+
+        result = mat_vec(vec, self.O, self.diag_shift)
+
+        if self.scale is not None:
+            result = jax.tree_multimap(jnp.multiply, result, self.scale)
+
+        # Reassemble real-imaginary split as needed
+        if self.mode != "holomorphic":
+            result = reassemble(result)
+
+        # Ravel PyTree back into vector as needed
+        if ravel:
+            result, _ = nkjax.tree_ravel(result)
+
+        return result
+
+    @jax.jit
+    def _unscaled_matmul(self, vec: PyTree) -> PyTree:
+        return mat_vec(vec, self.O, self.diag_shift)
+
+    @jax.jit
+    def _split_matmul(self, vec: Array) -> Array:
+        pars = self.params
+        if self.mode != "holomorphic":
+            pars = nkjax.tree_to_real(self.params)
+
+        _, unravel = nkjax.tree_ravel(pars)
+        vec = unravel(vec)
+
+        if self.scale is not None:
+            vec = jax.tree_multimap(jnp.multiply, vec, self.scale)
+
+        result = mat_vec(vec, self.O, self.diag_shift)
+
+        if self.scale is not None:
+            result = jax.tree_multimap(jnp.multiply, result, self.scale)
+
+        result, _ = nkjax.tree_ravel(result)
+        return result
+
+    @partial(jax.jit, static_argnums=1)
+    def _solve(self, solve_fun, y: PyTree, *, x0: Optional[PyTree] = None) -> PyTree:
+        """
+        Solve the linear system x=⟨S⟩⁻¹⟨y⟩ with the chosen iterataive solver.
+
+        Args:
+            y: the vector y in the system above.
+            x0: optional initial guess for the solution.
+
+        Returns:
+            x: the PyTree solving the system.
+            info: optional additional informations provided by the solver. Might be
+                None if there are no additional informations provided.
+        """
+        # Real-imaginary split RHS in R→R and R→C modes
+        if self.mode != "holomorphic":
+            y, reassemble = nkjax.tree_to_real(y)
+
+        if self.scale is not None:
+            y = jax.tree_multimap(jnp.divide, y, self.scale)
+            if x0 is not None:
+                x0 = jax.tree_multimap(jnp.multiply, x0, self.scale)
+
+        # to pass the object LinearOperator itself down
+        # but avoid rescaling, we pass down an object with
+        # scale = None
+        unsacaled_self = self.replace(scale=None)
+
+        out, info = solve_fun(unsacaled_self, y, x0=x0)
```

Comment:
1. Is this better than using `_unscaled_matmul`?
2. Does `self.replace` deep copy anything? If yes, let's not do this
3. Change `unsacaled_self` to `unscaled_self` please

Reply:
4. `unscaled_self.__matmul__` will want to disassemble the pytrees again. I don't think it should lead to errors, but it's wasteful

Reply:
```suggestion
        # scale factors are now removed, so we should only multiply with OᴴO+λ1
        out, info = solve_fun(self._unscaled_matmul, y, x0=x0)
```
it may be ugly, but it removes a nontrivial overhead

Reply:
Did you test if it actually does? I think it should not change anything...

Reply:
>Is this better than using _unscaled_matmul?

The point is that if I want to user solvers that require the full dense hamiltonian and diagonalize/svd it, we need to pass down to _solve the full hamiltonian.

>Does self.replace deep copy anything? If yes, let's not do this

No it does not. It duplicates the struct but the objects inside are not modified. And in any case this duplication will be compiled away by jax

> Change unsacaled_self to unscaled_self please

Indeed

Reply:
>unscaled_self.__matmul__ will want to disassemble the pytrees again. I don't think it should lead to errors, but it's wasteful

but it should get a dense array because we already disassembled it so it won't disassemble it, no?

Reply:
> > unscaled_self.**matmul** will want to disassemble the pytrees again. I don't think it should lead to errors, but it's wasteful
> 
> but it should get a dense array because we already disassembled it so it won't disassemble it, no?

I mean the real-imag splitting, it will try to look for complex leaves in the pytree after they've been removed once already.

Reply:
> > Is this better than using _unscaled_matmul?
> 
> The point is that if I want to user solvers that require the full dense hamiltonian and diagonalize/svd it, we need to pass down to _solve the full hamiltonian.

Fair enough

Reply:
But then we might want to remove `_unscaled_matmul` because it serves no other purpose (both here and in `QGTJacobianDense`

Reply:
> I mean the real-imag splitting, it will try to look for complex leaves in the pytree after they've been removed once already. Maybe this is a trivial overhead, I just have no sense of how expensive various kinds of jax magick are

So, things like `tree_cast` I think have no additional cost because jax is smart enough not to do those operations twice:
```
>>>import jax
>>>from functools import partial
>>>@partial(jax.jit, static_argnums=1)
>>>def test(x,i):
>>>	 return _test(x,i)

>>>def _test(x,i):
>>>  y = x.real
>>>  if i == 0:
>>>    return y
>>>  else:
>>>    return _test(y*y,i-1)

>>> jax.make_jaxpr(lambda x: test(x,2))(3.0+1j)
{ lambda  ; a.
  let b = xla_call[ backend=None
                    call_jaxpr={ lambda  ; a.
                                 let b = real a
                                     c = mul b b
                                     d = mul c c
                                 in (d,) }
                    device=None
                    donated_invars=(False,)
                    name=test ] a
  in (b,) }
```

About the splitting of real and complex part you are right.
That is why I now also set `mode=jholomphic` so to disable that logic. Maybe we could do it cleaner.